File: 100.same-tree.cpp
/*
 * @lc app=leetcode id=100 lang=cpp
 *
 * [100] Same Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        bool result = true;
        helper(p,q, result);
        return result;
    }

    void helper(TreeNode* p, TreeNode *q, bool& result){
        if (p != nullptr && q != nullptr){
            if (p->val != q->val)
                result = false;
            helper(p->left, q->left, result);
            helper(p->right, q->right, result);
        }
        else if (p == nullptr && q == nullptr)
            return ;
        else
            result = false;
    }
};
// @lc code=end

File: 100.same-tree.py
#
# @lc app=leetcode id=100 lang=python3
#
# [100] Same Tree
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def dfs(self, parent1: Optional[TreeNode], parent2: Optional[TreeNode]) -> bool:
        if parent1 is None and parent2 is None:
            return True
        if parent1 is None or parent2 is None:
            return False
        if parent1.val != parent2.val:
            return False
        return self.dfs(parent1.left, parent2.left) and self.dfs(parent1.right, parent2.right)
    
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        return self.dfs(p, q)
        
        
# @lc code=end

